### PR TITLE
Add context propagation sample

### PR DESCRIPTION
**Pull request by shreyassrivatsan**
- Simple context propagation sample showing ctx propagation through workflows and activities
- Change sample helpers to take in a context

### CODE REVIEW

File: cmd/samples/recipes/ctxpropagation/workflow.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+		HostID   string
+	}
+)
+
+// ApplicationName is the task list for this sample
+const ApplicationName = "CtxPropagatorGroup"
+
+// HostID - Use a new uuid just for demo so we can run 2 host specific activity workers on same machine.
+// In real world case, you would use a hostname or ip address as HostID.
+var HostID = ApplicationName + "_" + uuid.New()
```
Comment (mfateev) [inline cmd/samples/recipes/ctxpropagation/workflow.go]: May be call it ProcessID?

File: cmd/samples/recipes/ctxpropagation/README.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
+This sample workflow demos context propagation through a workflow.
+
+The workflow initializes the client with a context propagator which propagates
+across the workflow - available to use in workflow and activities.
```
Comment (mfateev) [inline cmd/samples/recipes/ctxpropagation/README.md]: I think it requires more info why it is needed or references to specific documents that explain that.

**Comment by dmetzgar** [comment]
Is this ready to merge?

File: cmd/samples/recipes/ctxpropagation/main.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+		MetricsScope:          h.Scope,
+		Logger:                h.Logger,
+		EnableLoggingInReplay: true,
+		ContextPropagators: []workflow.ContextPropagator{
+			NewContextPropagator(),
+		},
+	}
+	h.StartWorkers(h.Config.DomainName, ApplicationName, workerOptions)
+
+	// Host Specific activities processing case
```
Comment (samarabbas) [inline cmd/samples/recipes/ctxpropagation/main.go]: Do we really need it for this sample?

File: cmd/samples/recipes/ctxpropagation/main.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	"github.com/samarabbas/cadence-samples/cmd/samples/common"
+	"go.uber.org/cadence/client"
+	"go.uber.org/cadence/worker"
+	"go.uber.org/cadence/workflow"
+)
+
+// This needs to be done as part of a bootstrap step when the process starts.
+// The workers are supposed to be long running.
+func startWorkers(h *common.SampleHelper) {
+	// Configure worker options. Setup two context propagators - one string and
```
Comment (samarabbas) [inline cmd/samples/recipes/ctxpropagation/main.go]: Comment seems to be misleading.  I only see one ContextPropagator specified in the options below.

File: cmd/samples/recipes/ctxpropagation/workflow.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+		HeartbeatTimeout:       time.Second * 2, // such a short timeout to make sample fail over very fast
+	}
+	ctx = workflow.WithActivityOptions(ctx, ao)
+
+	if val := ctx.Value(propagateKey); val != nil {
+		vals := val.(Values)
+		workflow.GetLogger(ctx).Info("custom context propagated to workflow", zap.String(vals.Key, vals.Value))
+	}
+
+	var values map[string]string
```
Comment (samarabbas) [inline cmd/samples/recipes/ctxpropagation/workflow.go]: How is this set?  Activity only returns an error and nothing else.

File: cmd/samples/recipes/ctxpropagation/propagator.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+	"context"
+	"encoding/json"
+
+	"go.uber.org/cadence/workflow"
+)
+
+type (
+	contextKey struct{}
+
+	propagator struct{}
```
Comment (samarabbas) [inline cmd/samples/recipes/ctxpropagation/propagator.go]: Can you also include some documentation on this type.  Like what is the purpose, which interface it implements, translation of header properties -> context, etc.

**Comment by shreyassrivatsan** [comment]
> Is this ready to merge?

 @dmetzgar  Yes this is ready to merge now.

**Pull request merged**